### PR TITLE
Add vi solver option to phase field fracture

### PR DESCRIPTION
**Pull request by jiangwen84**
ref #14946

### CODE REVIEW

File: modules/tensor_mechanics/doc/content/source/materials/ComputeLinearElasticPFFractureStress.md [hunk 1]
```diff
@@ -229,6 +229,11 @@ Integration by parts yields
 The boundary integral term can be neglected as in most applications $d=0$ on $\partial\Omega$. Some authors [!cite](Mikelic2019) have proposed to replace the indicator function $d$ with $d^2$ in the first term in order to make the functional convex. The indicator function is implemented as a generic material object that can be easily provided and modified in an input file.
 The stress equilibrium and damage evolution equations are also modified to account for the pressure contribution.
 
+## PETSc SNES variational inequalities solver option
+
+Alternatively, the damage irreversibility condition can be enforced by using PETSc's SNES variational inequalities (VI) solver. In order to use PETSc's VI solver, [`PFFractureBoundsAux`](/phase_field/PFFractureBoundsAux.md) is required to set the bounds of phase field fracture variable. Note that in order for these bounds to have an effect, the user has to specify the
```
Comment (jiangwen84) [inline modules/tensor_mechanics/doc/content/source/materials/ComputeLinearElasticPFFractureStress.md]: @sapitts

Comment (sapitts) [reply modules/tensor_mechanics/doc/content/source/materials/ComputeLinearElasticPFFractureStress.md]: @jiangwen84 it looks like it is not possible to link from a phase field md file in a tensor mechanics file like you are trying to do here. While we do have some examples in the tensor mechanics documentation of linking to a phase field test, e.g.
```
moose/modules/tensor_mechanics/doc/content/source/materials/ComputeCrackedStress.md:!listing modules/combined/test/tests/phase_field_fracture/crack2d_computeCrackedStress_smallstrain.i
```
We don't have ability to link to a phase field md file, e.g.
```
moose/modules/tensor_mechanics/doc/content/modules/tensor_mechanics/index.md:and in combined physics simulations with the Heat Transfer, Phase Field, Contact, ....
```
I believe this limitation is by design to speed up the building of the MooseDocs, or is a function of how the MooseDocs uses the executable from each module.  In any case, the best option is to remove the link for the the `PFFractureBoundsAux` class.

Comment (jiangwen84) [reply modules/tensor_mechanics/doc/content/source/materials/ComputeLinearElasticPFFractureStress.md]: Thanks!

**Comment by jiangwen84** [comment]
@hugary1995 Could you take a look at it?

**Comment by jiangwen84** [comment]
@fdkong since this PR is related to PETSc VI solver, I feel that you should review and make a final merge decision. Thanks.

**Comment by hugary1995** [comment]
I have a few thoughts after taking a first look at this:
1. This auxkernel is almost too general to be called a PFFractureBoundsAux. Is it possible to derive from BoundsAux? I think it is better to have a base class for all bounds classes. Whether the auxkernel is defining a lower or upper bound should be specified as a parameter. Their can be a variety of ways of defining upper/lower bounds. Pretty much like we have ConstantIC and FunctionIC, we can also have ConstantBound, VariableOldBound (or call it Irreversibility), FunctionBound, etc.. The bounds for phase-field fracture is then just a combination of a ConstantBound (for upper bound) and a VariableOldBound (for lower bound).
2. Somewhere, perhaps in its constructor, we should check the solver parameters in the executioner to make sure that a VI solver is being used, otherwise we should issue an error or a warning.
3. In the stress calculator, we need to make sure only one of "history" and "VI" is used.

**Comment by jiangwen84** [comment]
@hugary1995 very good suggestions. I am waiting for @fdkong to commend on 1 and 2.

**Comment by jiangwen84** [comment]
I was thinking to make the Bounds more general, but I gave it up because variableOldBound seems to be quite specific to our phase field fracture model. Anyway, I am happy to rework on it if it turns out be more beneficial. For the second comment, I definitely want to add such check but I have not figured out how to yet.

File: modules/phase_field/src/auxkernels/PFFractureBoundsAux.C [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+PFFractureBoundsAux::PFFractureBoundsAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _upper_vector(_nl_sys.getVector("upper_bound")),
+    _lower_vector(_nl_sys.getVector("lower_bound")),
+    _bounded_var(_nl_sys.getVariable(_tid, getParam<NonlinearVariableName>("bounded_variable"))),
+    _bounded_var_name(parameters.get<NonlinearVariableName>("bounded_variable")),
+    _upper(getParam<Real>("upper"))
+{
+  if (!isNodal())
```
Comment (hugary1995) [inline modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: You can something like this here:

```cpp
  const std::vector<std::string> & solver_options =
      _app.getExecutioner()->getParamTempl<std::vector<std::string>>("petsc_options_value");
  if (std::find(solver_options.begin(), solver_options.end(), "vinewtonrsls") ==
          solver_options.end() &&
      std::find(solver_options.begin(), solver_options.end(), "vinewtonssls") ==
          solver_options.end())
    mooseError("A variational solver, i.e. vinewtonrsls, must be used in conjunction with PFFractureBoundsAux");
```

Comment (hugary1995) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: But this should be in the base class if you decide to derive it from BoundsAux. This also doesn't work with command line petsc options I guess.

Comment (fdkong) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: The actual right API is `petscOptionsDatabase`() in FEProblemBase, but it requires PETSc-3.12.x or higher.

It is better to add this check in the base class.

File: modules/phase_field/include/auxkernels/PFFractureBoundsAux.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+// Forward Declarations
+class PFFractureBoundsAux;
+
+template <>
+InputParameters validParams<PFFractureBoundsAux>();
+
+/**
+ * Provides the upper and lower bound of the phase field fracture variable
+ */
+class PFFractureBoundsAux : public AuxKernel
```
Comment (fdkong) [inline modules/phase_field/include/auxkernels/PFFractureBoundsAux.h]: You would derive from BoundAux.

File: modules/phase_field/doc/content/source/auxkernels/PFFractureBoundsAux.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+!syntax description /AuxKernels/PFFractureBoundsAux
+
+## Description
+
+`PFFractureBoundsAux` is used for setting bounds on phase field fracture variable. The upper bound is set as 1 by default. The lower bound is set as variable's old value in order to enforce irreversibility condition.
+
+Note that in order for these bounds to have an effect, the user has to specify the
+PETSc options `-snes_type vinewtonssls` or `-snes_type vinewtonrsls`. Both
+options emply semi-smooth non-linear solution algorithms to solve the
```
Comment (fdkong) [inline modules/phase_field/doc/content/source/auxkernels/PFFractureBoundsAux.md]: `vinewtonssls` is a semi-smooth Newton method, but `vinewtonrsls` does not employ the semi-smooth algorithm.  `vinewtonrsls` is the reduced space algorithm.

Comment (fdkong) [reply modules/phase_field/doc/content/source/auxkernels/PFFractureBoundsAux.md]: https://arxiv.org/pdf/math/0307305.pdf

File: modules/phase_field/doc/content/source/auxkernels/PFFractureBoundsAux.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+variational inequality. However, the `vinewtonrsls` option employs a reduced
+space active set method that actually reduces the size of the linear system by
```
Comment (fdkong) [inline modules/phase_field/doc/content/source/auxkernels/PFFractureBoundsAux.md]: `vinewtonrsls`  solves a reduced linear system of equations, corresponding to the inactive set, to calculate a direction. The solution regarding the inactive set will be updated based on the new direction, while the solution on the  active set will not be changed.

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
> * This auxkernel is almost too general to be called a PFFractureBoundsAux. Is it possible to derive from BoundsAux? I think it is better to have a base class for all bounds classes. Whether the auxkernel is defining a lower or upper bound should be specified as a parameter. Their can be a variety of ways of defining upper/lower bounds. Pretty much like we have ConstantIC and FunctionIC, we can also have ConstantBound, VariableOldBound (or call it Irreversibility), FunctionBound, etc.. The bounds for phase-field fracture is then just a combination of a ConstantBound (for upper bound) and a VariableOldBound (for lower bound).

Great suggestions, we really should do this unless it is unbelievably difficult to implement.

> * Somewhere, perhaps in its constructor, we should check the solver parameters in the executioner to make sure that a VI solver is being used, otherwise we should issue an error or a warning.

Yes!

File: modules/phase_field/src/auxkernels/PFFractureBoundsAux.C (hunk 2 above)
Comment (jiangwen84) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: How to make it backwards-compatible if PETSc version is lower than 3.12.x?

Comment (fdkong) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: The API `petscOptionsDatabase` includes options from both the input file and the command line. It requires 3.12.x or higher.

For old PETSc, you need to use the old way like:

(1) `PetscOptions & petsc = problem.getPetscOptions()` for the input file options

(2) `PetscGetArgs(&argc, &args);` for the comandline options.

PETSc-3.12.x will be the default solver within one month in MOOSE.

Comment (jiangwen84) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: Great.  Thanks!

Comment (fdkong) [reply modules/phase_field/src/auxkernels/PFFractureBoundsAux.C]: It may be easier and also better to add a routine `isSNESVI`() in `PetscSupport.C` . `PetscSupport.C` has some reference routines you could look at.

**Comment by jiangwen84** [comment]
@fdkong @hugary1995  Please take a second review.

File: framework/include/auxkernels/BoundsAuxBase.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+protected:
+  virtual Real computeValue() override;
+
+  /**
+   * Method to compute bound value of a variable.
+   */
+  virtual Real computeBound() = 0;
+
+  BoundType _type;
```
Comment (hugary1995) [inline framework/include/auxkernels/BoundsAuxBase.h]: Ben would ask you to add doxygen descriptions for all these members.

File: framework/include/auxkernels/ConstantBoundsAux.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  ConstantBoundsAux(const InputParameters & parameters);
+
+protected:
+  virtual Real computeBound() override { return _bound_value; }
+
+  Real _bound_value;
```
Comment (hugary1995) [inline framework/include/auxkernels/ConstantBoundsAux.h]: Doxygen description here - Ben

File: framework/src/utils/PetscSupport.C [hunk 8]
```diff
@@ -812,6 +812,40 @@ getCommonPetscKeys()
...
+  if (std::find(petsc.values.begin(), petsc.values.end(), "vinewtonssls") == petsc.values.end() &&
+      std::find(petsc.values.begin(), petsc.values.end(), "vinewtonrsls") == petsc.values.end() &&
+      std::find(cml_arg.begin(), cml_arg.end(), "vinewtonssls") == cml_arg.end() &&
+      std::find(cml_arg.begin(), cml_arg.end(), "vinewtonrsls") == cml_arg.end())
+    return false;
+
+  return true;
+
+#else
+  PetscOptions & petsc = fe_problem.petscOptionsDatabase();
```
Comment (jiangwen84) [inline framework/src/utils/PetscSupport.C]: @fdkong What is the correct way to use petscOptionsDatabase() for PETSc > 3.12? 

Got the following error and realized my usage could be wrong. 
/opt/civet/build_0/moose/framework/src/utils/PetscSupport.C:839:57: error: invalid initialization of reference of type 'Moose::PetscSupport::PetscOptions&' from expression of type 'PetscOptions' {aka '_n_PetscOptions*'}
  839 |   PetscOptions & petsc = fe_problem.petscOptionsDatabase();

File: framework/src/auxkernels/VariableOldValueBoundsAux.C [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+VariableOldValueBoundsAux::VariableOldValueBoundsAux(const InputParameters & parameters)
+  : BoundsAuxBase(parameters)
+{
+}
+
+Real
+VariableOldValueBoundsAux::computeBound()
+{
+  MooseVariable & var = _subproblem.getStandardVariable(_tid, _bounded_var_name);
```
Comment (hugary1995) [inline framework/src/auxkernels/VariableOldValueBoundsAux.C]: I don't know how much cost this is, but shouldn't this go into the initialization list? I wouldn't expect the variable to change during simulation.

File: modules/tensor_mechanics/src/materials/ComputePFFractureStressBase.C [hunk 10]
```diff
@@ -42,6 +46,7 @@ ComputePFFractureStressBase::ComputePFFractureStressBase(const InputParameters &
     _gc(getMaterialProperty<Real>("gc_prop")),
     _pressure(getDefaultMaterialProperty<Real>("fracture_pressure")),
     _use_current_hist(getParam<bool>("use_current_history_variable")),
+    _use_snes_vi_solver(getParam<bool>("use_snes_vi_solver")),
     _H(declareProperty<Real>("hist")),
     _H_old(getMaterialPropertyOld<Real>("hist")),
```
Comment (hugary1995) [inline modules/tensor_mechanics/src/materials/ComputePFFractureStressBase.C]: I think it makes more sense for _H_old to be a pointer. We really don't need _H_old if either _use_current_hist or _use_snes_vi_solver is true. But maybe this should go to another PR.

**Review by hugary1995 — COMMENTED** [review]
Looks good!

File: framework/src/utils/PetscSupport.C (hunk 8 above)
Comment (fdkong) [reply framework/src/utils/PetscSupport.C]: I think about a little more. I will uniform these interfaces in the future. I realized the old ways should work for PETSc-3.12.x as well.  Just use the old ways for everything at this point. I think it is too much for this PR to handle these dirty things.

**Comment by hugary1995** [comment]
Another possible issue: the default execute_on flags for auxkernels are "linear timestep_end". Is this appropriate?

**Comment by jiangwen84** [comment]
> Another possible issue: the default execute_on flags for auxkernels are "linear timestep_end". Is this appropriate?

@fdkong do you have answer for that?

**Comment by fdkong** [comment]
> > Another possible issue: the default execute_on flags for auxkernels are "linear timestep_end". Is this appropriate?
> 
> @fdkong do you have an answer for that?

`linear` should be good enough. I can not see the need to do `timestep_end`

File: framework/src/auxkernels/VariableOldValueBoundsAux.C (hunk 9 above)
Comment (jiangwen84) [reply framework/src/auxkernels/VariableOldValueBoundsAux.C]: Changed.

File: modules/tensor_mechanics/src/materials/ComputePFFractureStressBase.C (hunk 10 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/materials/ComputePFFractureStressBase.C]: Definitely another PR :-)

File: framework/src/utils/PetscSupport.C (hunk 8 above)
Comment (jiangwen84) [reply framework/src/utils/PetscSupport.C]: Changed.

**Comment by jiangwen84** [comment]
@fdkong Could you take another review?

**Review by fdkong — APPROVED** [review]

**Comment by permcody** [comment]
@jiangwen84 - This is breaking heavy modules tests. I imagine there's a easy fix?

**Comment by jiangwen84** [comment]
@permcody could you tell me where to look at the heavy modules tests results?

**Comment by fdkong** [comment]
https://civet.inl.gov/job/488767/
